### PR TITLE
Update for Loinc 2.68

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermLoaderSvcImpl.java [hunk 1]
```diff
@@ -119,7 +114,8 @@ public UploadStatistics loadLoinc(List<FileDescriptor> theFiles, RequestDetails
 				uploadProperties.getProperty(LOINC_IEEE_MEDICAL_DEVICE_CODE_MAPPING_TABLE_FILE.getCode(), LOINC_IEEE_MEDICAL_DEVICE_CODE_MAPPING_TABLE_FILE_DEFAULT.getCode()),
 				uploadProperties.getProperty(LOINC_IMAGING_DOCUMENT_CODES_FILE.getCode(), LOINC_IMAGING_DOCUMENT_CODES_FILE_DEFAULT.getCode()),
 				uploadProperties.getProperty(LOINC_PART_FILE.getCode(), LOINC_PART_FILE_DEFAULT.getCode()),
-				uploadProperties.getProperty(LOINC_PART_LINK_FILE.getCode(), LOINC_PART_LINK_FILE_DEFAULT.getCode()),
+				uploadProperties.getProperty(LOINC_PART_LINK_FILE_PRIMARY.getCode(), LOINC_PART_LINK_FILE_PRTIMARY_DEFAULT.getCode()),
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermLoaderSvcImpl.java:117]:
```suggestion
				uploadProperties.getProperty(LOINC_PART_LINK_FILE_PRIMARY.getCode(), LOINC_PART_LINK_FILE_PRIMARY_DEFAULT.getCode()),
```

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermLoaderSvcImpl.java [hunk 2]
```diff
@@ -469,6 +468,11 @@ UploadStatistics processLoincFiles(LoadedFileDescriptors theDescriptors, Request
 		handler = new LoincParentGroupFileHandler(code2concept, valueSets, conceptMaps, theUploadProperties);
 		iterateOverZipFile(theDescriptors, theUploadProperties.getProperty(LOINC_PARENT_GROUP_FILE.getCode(), LOINC_PARENT_GROUP_FILE_DEFAULT.getCode()), handler, ',', QuoteMode.NON_NUMERIC, false);
 
+		// Part link
+		handler = new LoincPartLinkHandler(codeSystemVersion, code2concept, propertyNamesToTypes);
+		iterateOverZipFile(theDescriptors, theUploadProperties.getProperty(LOINC_PART_LINK_FILE_PRIMARY.getCode(), LOINC_PART_LINK_FILE_PRTIMARY_DEFAULT.getCode()), handler, ',', QuoteMode.NON_NUMERIC, false);
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermLoaderSvcImpl.java:473]:
```suggestion
		iterateOverZipFile(theDescriptors, theUploadProperties.getProperty(LOINC_PART_LINK_FILE_PRIMARY.getCode(), LOINC_PART_LINK_FILE_PRIMARY_DEFAULT.getCode()), handler, ',', QuoteMode.NON_NUMERIC, false);
```

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincUploadPropertiesEnum.java [hunk 3]
```diff
@@ -67,9 +67,13 @@
 	// Part
 	LOINC_PART_FILE("loinc.part.file"),
 	LOINC_PART_FILE_DEFAULT("AccessoryFiles/PartFile/Part.csv"),
+
 	// Part link
-	LOINC_PART_LINK_FILE("loinc.part.link.file"),
-	LOINC_PART_LINK_FILE_DEFAULT("AccessoryFiles/PartFile/LoincPartLink.csv"),
+	LOINC_PART_LINK_FILE_PRIMARY("loinc.part.link.primary.file"),
+	LOINC_PART_LINK_FILE_PRTIMARY_DEFAULT("AccessoryFiles/PartFile/LoincPartLink_Primary.csv"),
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincUploadPropertiesEnum.java:73]:
```suggestion
	LOINC_PART_LINK_FILE_PRIMARY_DEFAULT("AccessoryFiles/PartFile/LoincPartLink_Primary.csv"),
```

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/TerminologyUploaderProviderR4Test.java [hunk 4]
```diff
@@ -492,7 +492,8 @@ private static void addFile(ZipOutputStream theZos, String theFileName) throws I
 		addFile(zos, LOINC_GROUP_FILE_DEFAULT.getCode());
 		addFile(zos, LOINC_GROUP_TERMS_FILE_DEFAULT.getCode());
 		addFile(zos, LOINC_PARENT_GROUP_FILE_DEFAULT.getCode());
-		addFile(zos, LOINC_PART_LINK_FILE_DEFAULT.getCode());
+		addFile(zos, LOINC_PART_LINK_FILE_PRTIMARY_DEFAULT.getCode());
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/TerminologyUploaderProviderR4Test.java:495]:
```suggestion
		addFile(zos, LOINC_PART_LINK_FILE_PRIMARY_DEFAULT.getCode());
```

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/TerminologyLoaderSvcLoincTest.java [hunk 5]
```diff
@@ -379,7 +379,8 @@ public static void addLoincMandatoryFilesToZip(ZipCollectionBuilder theFiles) th
 		theFiles.addFileZip("/loinc/", LOINC_ANSWERLIST_LINK_FILE_DEFAULT.getCode());
 		theFiles.addFileZip("/loinc/", LOINC_ANSWERLIST_LINK_DUPLICATE_FILE_DEFAULT.getCode());
 		theFiles.addFileZip("/loinc/", LOINC_PART_FILE_DEFAULT.getCode());
-		theFiles.addFileZip("/loinc/", LOINC_PART_LINK_FILE_DEFAULT.getCode());
+		theFiles.addFileZip("/loinc/", LOINC_PART_LINK_FILE_PRTIMARY_DEFAULT.getCode());
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/TerminologyLoaderSvcLoincTest.java:382]:
```suggestion
		theFiles.addFileZip("/loinc/", LOINC_PART_LINK_FILE_PRIMARY_DEFAULT.getCode());
```

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java [hunk 6]
```diff
@@ -86,44 +86,48 @@ public void accept(CSVRecord theRecord) {
...
-								concept.addPropertyCoding(nextPropertyName, ITermLoaderSvc.LOINC_URI, partNumber, nextPropertyValue);
-							} else {
-								String msg = "Unable to find part code with TYPE[" + key.getPartType() + "] and NAME[" + nextPropertyValue + "] (using name " + propertyValue + ")";
-								ourLog.warn(msg);
-//								throw new InternalErrorException(msg);
-							}
+							// These are handles by the LOINC PartLink file
 							break;
+
+							// FIXME: remove
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java:92]: FIXME needs to be addressed.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java [hunk 7]
```diff
@@ -86,44 +86,48 @@ public void accept(CSVRecord theRecord) {
...
+//							if (nextPropertyName.equals("COMPONENT")) {
+//								if (propertyValue.contains("^")) {
+//									propertyValue = propertyValue.substring(0, propertyValue.indexOf("^"));
+//								} else if (propertyValue.contains("/")) {
+//									propertyValue = propertyValue.substring(0, propertyValue.indexOf("/"));
+//								}
+//							}
+//
+//							PartTypeAndPartName key = new PartTypeAndPartName(nextPropertyName, propertyValue);
+//							String partNumber = myPartTypeAndPartNameToPartNumber.get(key);
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java:104]: I think you might be able to remove the `myPartTypeAndPartNameToPartNumber` property as this commented block of code appears to be the only thing referencing this property.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/TerminologyLoaderSvcLoincTest.java [hunk 8]
```diff
@@ -413,6 +414,9 @@ public void testLoadLoincMultiaxialHierarchySupport() throws Exception {
 		assertEquals(ITermLoaderSvc.LOINC_URI, code.getCodingProperties("PROPERTY").get(0).getSystem());
 		assertEquals("LP6802-5", code.getCodingProperties("PROPERTY").get(0).getCode());
 		assertEquals("Elpot", code.getCodingProperties("PROPERTY").get(0).getDisplay());
+		assertEquals(ITermLoaderSvc.LOINC_URI, code.getCodingProperties("PROPERTY").get(0).getSystem());
+		assertEquals("LP6802-5", code.getCodingProperties("PROPERTY").get(0).getCode());
+		assertEquals("Elpot", code.getCodingProperties("PROPERTY").get(0).getDisplay());
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/TerminologyLoaderSvcLoincTest.java:419]: I am confused here - the three new asserts here appear to simply duplicate the previous three assert statements. Were these supposed to be checking something else?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java (hunk 6 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java]: Oops- Done. Yup, will remove

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/TerminologyLoaderSvcLoincTest.java (hunk 8 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/TerminologyLoaderSvcLoincTest.java:421]: Removed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java (hunk 7 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java]: Oops- Done. Yup, will remove

**Pull request merged**